crates/hir-expand/src/lib.rs

```diff
@@ -225,6 +225,13 @@ pub enum MacroCallKind {
         derive_attr_index: AttrId,
         /// Index of the derive macro in the derive attribute
         derive_index: u32,
+        /// The "parent" macro call.
+        /// We will resolve the same token tree for all derive macros in the same derive attribute.
+        derive_macro_id: MacroCallId,
+    },
+    DeriveAttr {
+        ast_id: AstId<ast::Adt>,
+        invoc_attr_index: AttrId,
```

Comment:
We actually used to have this (as a field for `Attr`), but I got rid of that because it kind of infected the rest of the codebase with various checks which wasn't that great. Generally speaking there isn't really a reason for us to differentiate those two attributes. The only reason we'd need to here is to make our life easier wrt the IDE layer and derive attribute inputs. So I'd rather not we added this distinction back and instead figure out if we can explicitly encode this hack as a separate thing somehow (especially because there are other attributes, specifically builtin ones) where I was planning on employing this trick). I'll think about that a bit.

Reply:
I think separating also provides a higher understanding that these are different. Like we can see the expansion of the individual derive elements.

Reply:
I agree we should treat derive as a normal everyday attribute macro as much as possible. But, treating it specially  does two things.

We ensure the type of the syntax tree connected is an Adt. We also can do any other processing that could be required. Such as cfg stripping and stripping the `#[derive]`

I could undo this and at [MacroCallKind::Attr { ast_id, invoc_attr_index, .. }](https://github.com/wyatt-herkamp/rust-analyzer/blob/c376addfcc46f7eeec02d765027c8febc17a5825/crates/hir-expand/src/db.rs#L454-L455)
We would need to grab the attribute. Get it's name. Check if it is derive. If derive we would also have to pull the syntax tree and try to convert it to a an Adt. 

If that succeeds. We then can run the derive stripping and cfg stripping.

I personally thinks this adds complexity here. I think explicitly treating derive specially allows for us to have an easier experience when expanding them.

Reply:
That makes a lot of sense. I will get on it!